cpp/include/kvikio/file_handle.hpp

```diff
@@ -107,120 +126,121 @@ inline int open_fd(const std::string& file_path, const std::string& flags, mode_
-   * @param file_path File path to the file
-   * @param steal_fd When true, the handle owns the file descriptor and will close it
-   * on destruction. When false, the file is deregistered but not closed.
-   */
-  FileHandle(const std::string& file_path, const std::string& flags = "r", mode_t mode = m644)
-    : FileHandle(open_fd(file_path, flags, mode), true)
-  {
-  }
-
   /**
    * @brief FileHandle support move semantic but isn't copyable
    */
   FileHandle(const FileHandle&) = delete;
   FileHandle& operator=(FileHandle const&) = delete;
   FileHandle(FileHandle&& o) noexcept
-    : _fd{std::exchange(o._fd, -1)},
-      _own_fd{std::exchange(o._own_fd, false)},
-      _closed{std::exchange(o._closed, true)},
+    : _fd_direct_on{std::exchange(o._fd_direct_on, -1)},
+      _fd_direct_off{std::exchange(o._fd_direct_off, -1)},
+      _initialized{std::exchange(o._initialized, false)},
+      _compat_mode{std::exchange(o._compat_mode, false)},
       _nbytes{std::exchange(o._nbytes, 0)},
       _handle{std::exchange(o._handle, CUfileHandle_t{})}
   {
   }
   FileHandle& operator=(FileHandle&& o) noexcept
   {
-    _fd     = std::exchange(o._fd, -1);
-    _own_fd = std::exchange(o._own_fd, false);
-    _closed = std::exchange(o._closed, true);
-    _nbytes = std::exchange(o._nbytes, 0);
-    _handle = std::exchange(o._handle, CUfileHandle_t{});
+    _fd_direct_on  = std::exchange(o._fd_direct_on, -1);
+    _fd_direct_off = std::exchange(o._fd_direct_off, -1);
+    _initialized   = std::exchange(o._initialized, false);
+    _compat_mode   = std::exchange(o._compat_mode, false);
+    _nbytes        = std::exchange(o._nbytes, 0);
+    _handle        = std::exchange(o._handle, CUfileHandle_t{});
     return *this;
   }
+  ~FileHandle() noexcept { close(); }
 
-  /**
-   * @brief FileHandle support move semantic but isn't copyable
-   */
-  ~FileHandle() noexcept
-  {
-    if (!_closed) { this->close(); }
-  }
-
-  [[nodiscard]] bool closed() const noexcept { return _closed; }
+  [[nodiscard]] bool closed() const noexcept { return !_initialized; }
 
   /**
-   * @brief Deregister the file and close the file if created with `steal_fd=true`
+   * @brief Deregister the file and close the two files
    */
   void close() noexcept
   {
-    _closed = true;
+    if (closed()) { return; }
 #ifdef KVIKIO_CUFILE_EXIST
-    if (!defaults::compat_mode()) { cuFileAPI::instance()->HandleDeregister(_handle); }
+    if (!_compat_mode) { cuFileAPI::instance()->HandleDeregister(_handle); }
 #endif
-    if (_own_fd) { ::close(_fd); }
+    ::close(_fd_direct_on);
+    ::close(_fd_direct_off);
+    _fd_direct_on  = -1;
+    _fd_direct_off = -1;
+    _initialized   = false;
   }
 
   /**
-   * @brief Get the file descripter of the open file
+   * @brief Get the file descripter of one of the open files
+   *
    * @return File descripter
    */
-  [[nodiscard]] int fd() const noexcept { return _fd; }
+  [[nodiscard]] int fd() const noexcept { return _fd_direct_off; }
```

Comment:
Why use the one with `O_DIRECT` off?

Reply:
This is kind of an arbitrary decisions, my thinking is that `O_DIRECT`-off is the must compatible file descriptor.

Reply:
Should we document this then?

Reply:
I like to be a bit vague here as we might change this in the future. 
Change to:
```c++
  /**
   * @brief Get one of the file descriptors
   *
   * Notice, `FileHandle` maintains two file descriptors - one opened with the
   * `O_DIRECT` flag and one without. This function returns one of them but
   * it is unspecified with one.
   *
   * @return File descripter
   */
  [[nodiscard]] int fd() const noexcept { return _fd_direct_off; }

  /**
   * @brief Get the flags of one of the file descriptors (see open(2))
   *
   * Notice, `FileHandle` maintains two file descriptors - one opened with the
   * `O_DIRECT` flag and one without. This function returns the flags of one of
   * them but it is unspecified with one.
   *
   * @return File descripter
   */
  [[nodiscard]] int fd_open_flags() const { return open_flags(_fd_direct_on); }
```